src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentService.kt

```diff
@@ -290,4 +293,59 @@ class AppointmentService(
     }
     return appointmentDeliveryAddress
   }
+
+  fun scheduleNewAppointment(
+    sentReferralId: UUID,
+    appointmentType: AppointmentType,
+    durationInMinutes: Int,
+    appointmentTime: OffsetDateTime,
+    createdByUser: AuthUser,
+    appointmentDeliveryType: AppointmentDeliveryType,
+    appointmentSessionType: AppointmentSessionType,
+    appointmentDeliveryAddress: AddressDTO? = null,
+    npsOfficeCode: String? = null,
+  ): Appointment {
+    val sentReferral = referralRepository.findByIdAndSentAtIsNotNull(sentReferralId) ?: throw EntityNotFoundException(
+      "Sent Referral not found [referralId=$sentReferralId]"
+    )
+    val deliusAppointmentId =
+      communityAPIBookingService.book(sentReferral, null, appointmentTime, durationInMinutes, appointmentType, npsOfficeCode)
+    val appointment = Appointment(
+      id = UUID.randomUUID(),
+      appointmentTime = appointmentTime,
+      durationInMinutes = durationInMinutes,
+      deliusAppointmentId = deliusAppointmentId,
+      createdBy = authUserRepository.save(createdByUser),
+      createdAt = OffsetDateTime.now(),
+      referral = sentReferral,
+    )
+    appointmentRepository.saveAndFlush(appointment)
+    createOrUpdateAppointmentDeliveryDetails(appointment, appointmentDeliveryType, appointmentSessionType, appointmentDeliveryAddress, npsOfficeCode)
+    appointmentEventPublisher.appointmentScheduledEvent(appointment, appointmentType)
+    return appointment
+  }
+
+  // TODO: Add updatedBy/updatedAt to Appointment entity?
+  fun rescheduleExistingAppointment(
+    appointmentId: UUID,
+    appointmentType: AppointmentType,
+    durationInMinutes: Int,
+    appointmentTime: OffsetDateTime,
+    appointmentDeliveryType: AppointmentDeliveryType,
+    appointmentSessionType: AppointmentSessionType,
+    appointmentDeliveryAddress: AddressDTO? = null,
+    npsOfficeCode: String? = null,
+  ): Appointment {
+    val appointment = this.appointmentRepository.findById(appointmentId).orElseThrow { EntityNotFoundException("Appointment not found [appointmentId=$appointmentId]") }
+    appointment.appointmentFeedbackSubmittedAt ?.let { throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Appointment has already been delivered [appointmentId=$appointmentId]") }
```

Comment:
🙋‍♂️ Throwing HTTP specific response codes in a `Service` feels like overreaching -- if we wanted to use this method in other business logic, those non-controller functions end up dealing with http-related unchecked exceptions -- might be a relevant here to use checked exceptions?

(Also wondering if this fits better "The server cannot or will not process the request due to something that is perceived to be a client error" ([400 Bad Request](https://httpstatuses.com/400)) or "The request could not be completed due to a conflict with the current state of the target resource. This code is used in situations where the user might be able to resolve the conflict and resubmit the request" ([409 Conflict](https://httpstatuses.com/409)))

Reply:
Agreed. Now using EntityAlreadyExistsException and ValidationError for 409 and 400 respectively.